[AG and RS channel tuning] Add thread work threshold to tuning models and precompute reg index in LL128 (#1641)
ROCm/rccl: CHANGELOG.md

* Added new GPU target `gfx950`
* Added support for `unroll=1` in device-code generation to improve performance
* Added MSCCL support for AllGather single node and multinode (i.e., 8, 16 and 32 GPUs). To enable on multinode, set the
  environment variable `RCCL_MSCCL_FORCE_ENABLE=1`. Max message size for MSCCL AllGather usage is `12292 * sizeof(datatype) * nGPUs`.
* Thread thresholds for LL/LL128 are selected in Tuning Models for the MI300X. This impacts the number of channels used for AG and RS. Tuning model values are ignored if `NCCL_THREAD_THRESHOLDS` is set.
* Multi-node tuning for AllGather, AllReduce, and ReduceScatter that leverages LL/LL64/LL128 protocol to use nontemporal vector load/store for tunable message size ranges.
* LL/LL128 usage ranges for AR, AG, and RS are part of the tuning models, which enable architecture-specific tuning in conjunction with the existing Rome Models scheme in RCCL.

> * We expose two new APIs as part of an initiative to separate RCCL code. These APIs are `rcclGetAlgoInfo` and `rcclFuncMaxSendRecvCount`. However, user-level invocation requires that RCCL is built with `RCCL_EXPOSE_STATIC` enabled.

Reviewer: ```suggestion
* Two new APIs are exposed as part of an initiative to separate RCCL code. These APIs are `rcclGetAlgoInfo` and `rcclFuncMaxSendRecvCount`. However, user-level invocation requires that RCCL be built with `RCCL_EXPOSE_STATIC` enabled.
```

> * Two new APIs are exposed as part of an initiative to separate RCCL code. These APIs are `rcclGetAlgoInfo` and `rcclFuncMaxSendRecvCount`. However, user-level invocation requires that RCCL be built with `RCCL_EXPOSE_STATIC` enabled.

Reviewer 2: What is "RCCL_EXPOSE_STATIC", how to build rccl with that enabled?

Author: `RCCL_EXPOSE_STATIC` is a compilation flag to make it possible to access C-style static functions outside of the source file where they are defined in RCCL/NCCL. Functions can be gradually added and explored by using `rccl_static` instead of `static` keyword. 
It is unavailable through `install.sh`, but can be added manually to a CMake command. An example use is through ./tools/topo_expl/Makefile

* Setting `NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=VERBS` will generate traces for fifo and data ibv_post_sends
* Added `--log-trace` flag to enable traces through the install.sh script (e.g. `./install.sh --log-trace`)
* Added MSCCL support for AllGather single node and multinode (i.e., 8, 16 and 32 GPUs). To enable on multinode, set the

>   environment variable `RCCL_MSCCL_FORCE_ENABLE=1`. Max message size for MSCCL AllGather usage is 12292 * sizeof(datatype) * nGPUs
> * Thread thresholds for LL/LL128 is selected in Tuning Models for MI300x. It currently impacts the number of channels used for AG and RS. Tuning model values are ignored if `NCCL_THREAD_THRESHOLDS` is set

Reviewer: ```suggestion
  environment variable `RCCL_MSCCL_FORCE_ENABLE=1`. Max message size for MSCCL AllGather usage is `12292 * sizeof(datatype) * nGPUs`.
* Thread thresholds for LL/LL128 are selected in Tuning Models for the MI300X. This impacts the number of channels used for AG and RS. Tuning model values are ignored if `NCCL_THREAD_THRESHOLDS` is set.
```
